### PR TITLE
Implement VK_KHR_depth_stencil_resolve

### PR DESCRIPTION
**Pull request by mark-lunarg**
Tested with `vulkan_test_applications::depth_stencil_resolve.apk`.

### CODE REVIEW

File: gapis/api/vulkan/api/enums.api [hunk 1]
```diff
@@ -375,9 +375,15 @@ enum VkStructureType: u32 {
   VK_STRUCTURE_TYPE_ATTACHMENT_REFERENCE_STENCIL_LAYOUT_KHR                     = 1000241001,
   VK_STRUCTURE_TYPE_ATTACHMENT_DESCRIPTION_STENCIL_LAYOUT_KHR                   = 1000241002,
 
+  // @extension("VK_KHR_depth_stencil_resolve")
+  VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_DEPTH_STENCIL_RESOLVE_PROPERTIES_KHR = 1000199000,
```
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/api/enums.api:379]: We don't need to repeat this as this is a core extension. We shall only use the core version, as the values are same.

File: gapis/api/vulkan/api/enums.api [hunk 2]
```diff
@@ -1179,3 +1185,19 @@ enum VkTessellationDomainOrigin: u32  {
...
+// Vulkan 1.2 core
+enum VkResolveModeFlagBits: u32 {
+  VK_RESOLVE_MODE_NONE                = 0x00000000,
+  VK_RESOLVE_MODE_SAMPLE_ZERO_BIT     = 0x00000001,
+  VK_RESOLVE_MODE_AVERAGE_BIT         = 0x00000002,
+  VK_RESOLVE_MODE_MIN_BIT             = 0x00000004,
+  VK_RESOLVE_MODE_MAX_BIT             = 0x00000008,
+
+  //@extension("VK_KHR_depth_stencil_resolve")
+  VK_RESOLVE_MODE_NONE_KHR            = 0x00000000,
```
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/api/enums.api:1198]: Same with above

File: gapis/api/vulkan/state_rebuilder.go [hunk 3]
```diff
@@ -2441,6 +2441,31 @@ func (sb *stateBuilder) createRenderPass2(rp RenderPassObjectʳ) {
...
 		sd := rp.SubpassDescriptions().Get(k)
+		sdPnext := NewVoidᶜᵖ(memory.Nullptr)
+
+		if !sd.DepthStencilResolve().IsNil() {
+			depthStencilResolve := sd.DepthStencilResolve().Get()
+
+			dsrAttachmentRef := NewVkAttachmentReference2ᶜᵖ(sb.MustAllocReadData(
+				NewVkAttachmentReference2(
+					VkStructureType_VK_STRUCTURE_TYPE_ATTACHMENT_REFERENCE_2,
+					NewVoidᶜᵖ(memory.Nullptr),
```
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/state_rebuilder.go:2452]: AttachmentReference2 here can originally have pNext that we are missing currently.

File: gapis/api/vulkan/api/renderpass_framebuffer.api [hunk 4]
```diff
@@ -405,6 +412,29 @@ sub ref!RenderPassObject createRenderPassObjectFromInfo2(
...
+        sType := as!const VkStructureType*(next.Ptr)[0]
+        switch sType {
+          case VK_STRUCTURE_TYPE_SUBPASS_DESCRIPTION_DEPTH_STENCIL_RESOLVE: {
+            ext := as!VkSubpassDescriptionDepthStencilResolve*(next.Ptr)[0]
+            description.DepthStencilResolve = new!SubpassDescriptionDepthStencilResolve(
+              DepthResolveMode: ext.depthResolveMode,
+              StencilResolveMode: ext.stencilResolveMode,
+            )
+            attachment := ext.pDepthStencilResolveAttachment[0]
+            description.DepthStencilResolve.DepthStencilResolveAttachment = AttachmentReference(
```
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/api/renderpass_framebuffer.api:428]: We are missing pNext handling here.

**Review by yalcinmelihyasin — CHANGES_REQUESTED** [review]

**Comment by yalcinmelihyasin** [comment]
There are some minor comments about adding KHR at the end but I believe that this PR shown a fundamental flow in our API files in regards to pNext handling.

We handle pNext inline in the code without proper modularization.  Therefore, the new sub structures in Vulkan e.g VkAttachmentReference2 that can be included in many places now is pain to handle. I would like to hold this PR a bit and talk about how we can restructure this.

File: gapis/api/vulkan/api/renderpass_framebuffer.api (hunk 4 above)
Comment (mark-lunarg) [reply gapis/api/vulkan/api/renderpass_framebuffer.api:428]: @yalcinmelihyasin, just to be clear, the other `AttachmentReference()` usages on lines 442, 453, 465 and 476 also need 'pNext' handling, yes?

Comment (yalcinmelihyasin) [reply gapis/api/vulkan/api/renderpass_framebuffer.api:428]: Yes. with the new PR merging, we are able to do this now.

File: gapis/api/vulkan/api/enums.api (hunk 1 above)
Comment (mark-lunarg) [reply gapis/api/vulkan/api/enums.api]: Done.

File: gapis/api/vulkan/api/enums.api (hunk 2 above)
Comment (mark-lunarg) [reply gapis/api/vulkan/api/enums.api]: Done.

File: gapis/api/vulkan/state_rebuilder.go (hunk 3 above)
Comment (mark-lunarg) [reply gapis/api/vulkan/state_rebuilder.go]: Done, please check this.  Test appears to confirm it is correct.

**Review by yalcinmelihyasin — APPROVED** [review]
LGTM

**Pull request merged**